lib/finch/http1/pool.ex

```diff
@@ -38,6 +38,112 @@ defmodule Finch.HTTP1.Pool do
     )
   end
 
+  @impl Finch.Pool
+  def stream(pool, req, name, opts) do
+    pool_timeout = Keyword.get(opts, :pool_timeout, 5_000)
+    receive_timeout = Keyword.get(opts, :receive_timeout, 15_000)
+    request_timeout = Keyword.get(opts, :request_timeout, 30_000)
+
+    metadata = %{request: req, pool: pool, name: name}
+
+    start_time = Telemetry.start(:queue, metadata)
+    owner = self()
+    ref = make_ref()
+
+    holder =
+      spawn_link(fn ->
+        try do
+          NimblePool.checkout!(
+            pool,
+            :checkout,
+            fn from, {state, conn, idle_time} ->
+              Telemetry.stop(:queue, start_time, metadata, %{idle_time: idle_time})
+              send(owner, {ref, :ok, {conn, from, state, idle_time}})
+
+              receive do
+                {^ref, :stop, state} -> {:ok, state}
+              end
+            end,
+            pool_timeout
+          )
+        catch
+          :exit, data ->
+            Telemetry.exception(:queue, start_time, :exit, data, __STACKTRACE__, metadata)
+            send(owner, {ref, :exit, {data, __STACKTRACE__}})
+        end
+      end)
+
+    receive do
+      {^ref, :ok, conn_from} ->
+        {conn, from, state, idle_time} = conn_from
+
+        with {:ok, conn} <- Conn.connect(conn, name),
+             {:ok, conn} <- transfer_if_open(conn, state, from),
+             {:ok, stream} <-
+               Conn.stream(
+                 conn,
+                 req,
+                 name,
+                 ref,
+                 holder,
+                 receive_timeout,
+                 request_timeout,
+                 idle_time
+               ) do
+          {:ok, stream}
+        else
+          :closed ->
+            send(holder, {ref, :stop, state})
+            # FIXME
+            {:error, :closed}
+
+          {:error, conn, error} ->
+            state =
+              if Conn.open?(conn) do
+                {:ok, conn}
+              else
+                :closed
+              end
+
+            send(holder, {ref, :stop, state})
+            {:error, error}
+        end
+
+      {^ref, :error, reason} ->
+        {:error, reason}
+
+      {^ref, :exit, data_trace} ->
+        {data, trace} = data_trace
+
+        case data do
+          {:timeout, {NimblePool, :checkout, _affected_pids}} ->
+            # Provide helpful error messages for known errors
+            reraise(
+              """
+              Finch was unable to provide a connection within the timeout due to excess queuing \
+              for connections. Consider adjusting the pool size, count, timeout or reducing the \
+              rate of requests if it is possible that the downstream service is unable to keep up \
+              with the current rate.
+              """,
+              trace
+            )
+
+          _ ->
+            exit(data)
+        end
```

Comment:
There is some code duplication with other functions, those should be extracted out.

Reply:
I agree, but I think that DRYing should be done after everyone agress on implementation detail